### PR TITLE
Add support for Common I/O (UART, SPI and I2C) on ESP32

### PR DESCRIPTION
**Pull request by shubhamkulkarni97**
Description
-----------
This PR intends to add support for UART, SPI and I2C modules in Common I/O

Checklist:
----------

- [x] I have tested my changes. No regression in existing tests.
- [x] My code is Linted.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Comment by qiutongs** [comment]
As discussed, ESP is updating the implementation for async function. Using FreeRTOS timer is not the preferable option.

**Review by YTvW — DISMISSED** [review]

**Comment by teuteuguy** [comment]
Any updates on this PR ?

**Comment by lundinc2** [comment]
@shubhamkulkarni97 @teuteuguy,

As discussed originally, we were going to wait for Async, without the reliance of FreeRTOS timer. We will reverse that decision and take this as is with the hope of adding that later.

@shubhamkulkarni97 I believe to get this merged, all we need to do is rebase this branch, and resolve two variables causing the build failures. 

The issue is caused by `uctestIotI2CSlaveAddr` and `ucAssistedTestIotI2CSlaveAddr`. Alongside this, I will remove the calls to `portYIELD_FROM_ISR` that are generating build errors when building for the esp32.

Please let me know if you are still willing to do this, otherwise I can create a mirror'd PR with those changes.

Thanks,

Carl

**Comment by shubhamkulkarni97** [comment]
@teuteuguy @lundinc2, will rebase PR to latest master and make it available for review soon.

**Comment by shubhamkulkarni97** [comment]
Hi @lundinc2 @teuteuguy,

I have updated the PR with latest changes. With this change, common-io drivers don't have any dependency on FreeRTOS timer for async implementation.

There is one failure in UART test case(`AFQP_IotUARTWriteAsyncReadAsyncLoopbackTest`).
I believe it is a bug in test case as there is no delay between start of transaction and request for [eGetTxNoOfbytes](https://github.com/aws/amazon-freertos/blob/dd97ccc7e621d8de8cf579854f4672c92e597c13/libraries/abstractions/common_io/test/test_iot_uart.c#L404).

Moving [vTaskDelay](https://github.com/aws/amazon-freertos/blob/dd97ccc7e621d8de8cf579854f4672c92e597c13/libraries/abstractions/common_io/test/test_iot_uart.c#L409) above request for `eGetTxNoOfbytes` fixes the failure.

Please help verify this issue.

Thanks,
Shubham

**Comment by lundinc2** [comment]
Hi @shubhamkulkarni97, I agree this is a bug. I will update this.

**Comment by shubhamkulkarni97** [comment]
License for newly added files brought in sync with other Espressif port files.

**Review by lundinc2 — APPROVED** [review]

**Review by cobusve — APPROVED** [review]

File: vendors/espressif/boards/esp32/CMakeLists.txt [hunk 1]
```diff
@@ -324,6 +325,34 @@ target_link_libraries(
...
+)
+target_include_directories(
+    AFR::common_io::mcu_port
+    INTERFACE
+        "${afr_ports_dir}/common_io/include"
+        "${AFR_ROOT_DIR}/libraries/abstractions/common_io/test"
+)
+
+
+target_compile_options(
```
Comment (sarenameas) [inline vendors/espressif/boards/esp32/CMakeLists.txt:346]: Why did we add more compile options?

File: vendors/espressif/boards/esp32/CMakeLists.txt [hunk 2]
```diff
@@ -349,6 +378,7 @@ target_link_libraries(
         AFR::pkcs11
         AFR::ota_mqtt
         AFR::ota_http
+        AFR::common_io
```
Comment (sarenameas) [inline vendors/espressif/boards/esp32/CMakeLists.txt:381]: Does OTA need common_io?

File: vendors/espressif/boards/esp32/CMakeLists.txt (hunk 1 above)
Comment (shubhamkulkarni97) [reply vendors/espressif/boards/esp32/CMakeLists.txt:346]: There were a bunch of compile errors when common-io is enabled. These compile options fix those errors.

File: vendors/espressif/boards/esp32/CMakeLists.txt (hunk 2 above)
Comment (shubhamkulkarni97) [reply vendors/espressif/boards/esp32/CMakeLists.txt:381]: OTA does not need common-io and this change is added by mistake. Actually this change was added when @lundinc2 created a PR to enable common-io on esp32 in my personal fork: https://github.com/shubhamkulkarni97/amazon-freertos/pull/1/files

Sorry for overlooking this change.

**Pull request merged**